### PR TITLE
[CRB-155][CRB-158] Request settings in the proper context, remove background update thread

### PR DESCRIPTION
**Pull request by nathanwhit**
This PR removes the background `SettingsUpdater` and shifts to requesting on-chain settings as needed by transactions. We now request the state through the appropriate channels, which means we are now subject to permission checking (resolving CRB-155) and the values of settings are fully deterministic (resolving CRB-158). 
This incurs one extra state request on transactions which depend on the value of settings. This also simplifies the code greatly and removes the last piece of mutable global state in the processor.

### CODE REVIEW

**Review by CAGS295 — APPROVED** [review]

**Review by CAGS295 — APPROVED** [review]

**Comment by CAGS295** [comment]
@will-gluwa merged into main?

**Comment by CAGS295** [comment]
How are state changes visible from the context's PoV. Is it like a view from the previous block? Or it depends on the scheduler? Are state changes from previous batches within the scheduler already visible to subsequent scheduled txns?

**Comment by CAGS295** [comment]
Now that the update loop has been removed, are settings being updated on demand at all places where it is needed?

File: src/handler/context.rs [hunk 1]
```diff
@@ -60,8 +90,41 @@ impl HandlerContext {
...
+                if state.is_empty() {
+                    log::debug!("setting not found for key {:?}", key);
+                    Ok(None)
+                } else {
+                    let (_addr, value) = &state[0];
+                    Self::find_setting(&value, key)
+                }
+            }
+            Err(e) => Err(e.into()),
+        }
```
Comment (atodorov) [inline src/handler/context.rs:127]: These 2 functions can use some unit testing love.  Other than that I'm not sure what else needs testing.

**Comment by nathanwhit** [comment]
> 
> 
> How are state changes visible from the context's PoV. Is it like a view from the previous block? Or it depends on the scheduler? Are state changes from previous batches within the scheduler already visible to subsequent scheduled txns? What I'm trying to figure out is if settings are effective immediately with in a block or not.

Long answer:

The context is initialized with a view based on a specific state root hash. The only state that is meant to be accessed by the TP is the state rooted at that hash. Before we were essentially creating a context with an empty state root hash, which means it accesses the global state now, regardless of whether it *should* be visible to us or not.

The actual state root hash is provided by the scheduler. When we validate a block:

1. We create a new scheduler with the state root from the previous block (the state tree resulting from the previous block, that is)
2. We add all batches in the block to the scheduler
3. The scheduler creates a predecessor tree based on the input/output declarations for each transaction (assuming the inputs/outputs _do_ conflict, I believe the relative ordering of transactions depends on the order in which they appear, though that is not guaranteed). The predecessor tree doesn't distinguish between batches - all of the transactions from all batches are put in the same tree.
4. The scheduler continually pulls off the next transaction that doesn't have any pending predecessors (or blocks until one as available).
5. The result of executing a transaction is visible to all "descendants" i.e. if transaction A is a predecessor of B, then B will see the state resulting from the execution of A

Short answer:

> Is it like a view from the previous block?

Only for the first transaction in a given line of predecessors. If txn A B C D are executed in that order, A will see the state from the previous block, but B will see the state from the execution of A, and so on

> Or it depends on the scheduler

Yes pretty much it depends on the scheduler. The exact order that transactions are executed is unspecified.

> Are state changes from previous batches within the scheduler already visible to subsequent scheduled txns?

Yes, if the state changes were a result of a predecessor transaction.

> What I'm trying to figure out is if settings are effective immediately with in a block or not.

Yes, they are effective immediately.

**Comment by nathanwhit** [comment]
> 
> 
> Now that the update loop has been removed, are settings being updated on demand at all places where it is needed?

Yes they are fetched on demand whenever needed. The only setting that would be read more than once within a transaction is the fee (in #7), so in that PR I cache the fee for the remainder of the transaction to prevent unnecessary state requests.

**Pull request merged**